### PR TITLE
Add support for project level deployer role [SAMSON-169]

### PR DESCRIPTION
**Pull request by sbrnunes**
Adding support to allow an User with project role 'Deployer' to have deployer access to a project regardless of his system level role.

/cc @zendesk/samson 

### References
- [SAMSON-169](https://zendesk.atlassian.net/browse/SAMSON-169)

### Risks
- Changes to the behaviour of authorisation filters (confirm no security holes were added)

### CODE REVIEW

File: app/controllers/locks_controller.rb [hunk 1]
```diff
@@ -1,32 +1,51 @@
...
+  end
+
   before_action :authorize_admin!, if: :for_global_lock?
 
   def create
     attributes = params.require(:lock).
-      permit(:description, :stage_id, :warning).
-      merge(user: current_user)
+        permit(:description, :stage_id, :warning).
+        merge(user: current_user)
```
Comment (henders) [inline app/controllers/locks_controller.rb]: indent too big! :)

File: app/controllers/locks_controller.rb [hunk 2]
```diff
@@ -1,32 +1,51 @@
...
   protected
 
   def for_global_lock?
     case action_name
-    when "create" then !params[:lock].try(:[], :stage_id)
-    when "destroy" then !lock.stage_id
-    else
-      raise "Unsupported action"
+      when 'create' then
+        !params[:lock].try(:[], :stage_id) || params[:lock].try(:[], :stage_id).try(:empty?)
```
Comment (henders) [inline app/controllers/locks_controller.rb]: Why did this line have to change? if params[:lock][:stage_id] is potentially an empty string then
```!params[:lock].try(:[], :stage_id).presence``` should cover it

Comment (henders) [inline app/controllers/locks_controller.rb]: also, indent less :)

File: app/models/project.rb [hunk 3]
```diff
@@ -125,6 +125,20 @@ def last_deploy_by_group(before_time)
...
   end
 
+  def self.search_by_criteria(criteria)
+    scope = Project
+    scope = scope.search(criteria[:search]) if criteria[:search]
+    scope.order("#{sort_column(criteria[:sort])} #{sort_direction(criteria[:direction])}").page(criteria[:page])
+  end
+
+  def self.sort_column(column)
+    column_names.include?(column) ? column : "created_at"
```
Comment (henders) [inline app/models/project.rb]: Not sure if hiding bad input is the way to go... if user wanted to sort by a field they typo'd they should probably get an error.

File: app/models/user.rb [hunk 4]
```diff
@@ -60,18 +60,32 @@ def gravatar_url
...
   end
 
+  def self.search_by_criteria(criteria)
+    scope = User
+    scope = scope.search(criteria[:search]) if criteria[:search]
+    scope.order("#{sort_column(criteria[:sort])} #{sort_direction(criteria[:direction])}").page(criteria[:page])
+  end
+
+  def self.sort_column(column)
+    column_names.include?(column) ? column : "created_at"
```
Comment (henders) [inline app/models/user.rb]: Same as above for Project

File: test/controllers/builds_controller_test.rb [hunk 5]
```diff
@@ -98,4 +98,85 @@ def stub_git_reference_check(returns: false)
       end
     end
   end
+
+  as_a_viewer_project_deployer do
```
Comment (henders) [inline test/controllers/builds_controller_test.rb]: now that's a confusing name :)

File: app/controllers/locks_controller.rb (hunk 2 above)
Comment (sbrnunes) [inline app/controllers/locks_controller.rb]: This one was hard to remember why did I changed it. :-P
But yes, I noticed that `stage_id` is always present in the parameters, and when to provided, it comes as an empty string.

I was not aware of the `presence` method. Looks much better, thanks. I'll check it out.

File: app/models/project.rb (hunk 3 above)
Comment (sbrnunes) [inline app/models/project.rb]: So, would something like this be better:
```
def self.sort_column(column)
  column.presence || "created_at"
end
```

By default, when not provided, we need to sort by created_at.

File: test/controllers/builds_controller_test.rb (hunk 5 above)
Comment (sbrnunes) [inline test/controllers/builds_controller_test.rb]: Oh man, I know, couldn't come up with a better name. :)

Here I'd like to make it clear that this user is a User with 'Viewer' system level role, but he also has the 'Deployer' role on a specific project.

Would you prefer just `as_a_project_deployer`? Can you remember of a better name for this?

File: app/controllers/locks_controller.rb (hunk 1 above)
Comment (sbrnunes) [inline app/controllers/locks_controller.rb]: Fixed. Damn RubyMine code styles. :)

File: app/models/project.rb [hunk 6]
```diff
@@ -126,6 +126,20 @@ def last_deploy_by_group(before_time)
     releases.map { |group_id, deploys| [ group_id, deploys.sort_by(&:updated_at).last ] }.to_h
   end
 
+  def self.search_by_criteria(criteria)
+    scope = Project
+    scope = scope.search(criteria[:search]) if criteria[:search]
+    scope.order("#{sort_column(criteria[:sort])} #{sort_direction(criteria[:direction])}").page(criteria[:page])
```
Comment (steved) [inline app/models/project.rb]: without the strict check against format, this is injection waiting to happen

File: app/models/user.rb [hunk 7]
```diff
@@ -60,17 +60,31 @@ def gravatar_url
...
 
   def project_role_for(project)
     user_project_roles.find_by(project: project)
   end
 
+  def self.search_by_criteria(criteria)
+    scope = User
+    scope = scope.search(criteria[:search]) if criteria[:search]
+    scope.order("#{sort_column(criteria[:sort])} #{sort_direction(criteria[:direction])}").page(criteria[:page])
+  end
```
Comment (steved) [inline app/models/user.rb]: This probably could be in a concern `Searchable` for user and project

File: app/controllers/concerns/current_project.rb [hunk 8]
```diff
@@ -12,6 +12,7 @@ def current_project
   protected
 
   def find_project(param)
+    # Will this return not found if project does not exist ?
```
Comment (steved) [inline app/controllers/concerns/current_project.rb]: :scissors:

File: app/controllers/locks_controller.rb [hunk 9]
```diff
@@ -7,26 +13,39 @@ def create
...
   end
 
   protected
 
   def for_global_lock?
     case action_name
-    when "create" then !params[:lock].try(:[], :stage_id)
-    when "destroy" then !lock.stage_id
+    when 'create' then
+      !params[:lock].try(:[], :stage_id) || !params[:lock].try(:[], :stage_id).presence
```
Comment (steved) [inline app/controllers/locks_controller.rb]: presence includes nil, so you just need the second check

File: test/controllers/locks_controller_test.rb [hunk 10]
```diff
@@ -1,21 +1,37 @@
...
   let(:global_lock) { Lock.create! user: users(:deployer) }
 
   before { request.headers['HTTP_REFERER'] = '/back' }
 
   as_a_viewer do
     unauthorized :post, :create
-    unauthorized :post, :create, lock: {stage_id: 1}
-    unauthorized :delete, :destroy, id: 1
+
+    it('responds with unauthorized when doing a post to create a local lock') do
```
Comment (steved) [inline test/controllers/locks_controller_test.rb]: parens are inconsistent with other tests, only use them for single lines

File: test/controllers/locks_controller_test.rb [hunk 11]
```diff
@@ -82,4 +98,51 @@
       end
     end
   end
+
+  as_a_project_deployer do
+    unauthorized :post, :create
+
+    it("responds with unauthorized when doing a post to create a global lock") { post :create, id: global_lock.id }
```
Comment (steved) [inline test/controllers/locks_controller_test.rb]: way too long for a single line and it doesn't test anything

File: app/models/project.rb (hunk 6 above)
Comment (sbrnunes) [inline app/models/project.rb]: Damn, you're right, I missed that.

File: app/controllers/concerns/current_project.rb (hunk 8 above)
Comment (sbrnunes) [inline app/controllers/concerns/current_project.rb]: Removed in another PR.

File: app/controllers/locks_controller.rb (hunk 9 above)
Comment (sbrnunes) [inline app/controllers/locks_controller.rb]: I'm not worried with the nil, I'm worried with the `stage_id` being sent as an empty string even for a global lock, which I noticed was happening and the tests were not covering it.

What about: `!params[:lock].try(:[], :stage_id) || !params[:lock][:stage_id].presence`

This should return true if:
- params does not contain `stage_id` (first check)
- params contains `stage_id` but `stage_id` is an empty string (second check)

Comment (steved) [inline app/controllers/locks_controller.rb]: Right. `params[:lock].try(:[], :stage_id).present?` will cover both of those cases.

File: test/controllers/locks_controller_test.rb (hunk 11 above)
Comment (sbrnunes) [inline test/controllers/locks_controller_test.rb]: Indeed. Something went really weird with this line.

Fixed.

**Pull request closed without merging**